### PR TITLE
OIDC: track tokens separately to devices

### PR DESCRIPTION
**Pull request by DMRobertson**
This is the first part of my plan proposed in https://github.com/matrix-org/sliding-sync/issues/89#issuecomment-1527909808. Note that it targets a separate `dmr/oidc` branch.

This diff

- introduces a v2 tokens table,
- changes the schema of the v2 devices table,
- changes the sync2 part of the application to uses the tables, and
- passes (user, device) pairs around as conn and poller IDs, including to the sync3 part of the application

It does not
- alter the v3 parts of the schema (want to use actual device IDs and to include user IDs), nor
- add any sort of migration.

### CODE REVIEW

File: pubsub/v2.go [hunk 1]
```diff
@@ -78,6 +78,7 @@ type V2InitialSyncComplete struct {
 func (*V2InitialSyncComplete) Type() string { return "V2InitialSyncComplete" }
 
 type V2DeviceData struct {
+	UserID   string
 	DeviceID string
 	Pos      int64
```
Comment (DMRobertson) [inline pubsub/v2.go:83]: Is this actually Pos or is it a since token?

Comment (DMRobertson) [reply pubsub/v2.go:83]: It's a position in the `syncv3_device_data_seq` sequence.

File: sync2/handler2/handler.go [hunk 2]
```diff
@@ -150,12 +154,11 @@ func (h *Handler) OnTerminated(userID, deviceID string) {
...
-	h.Store.ToDeviceTable.DeleteAllMessagesForDevice(deviceID)
-	h.Store.DeviceDataTable.DeleteDevice(userID, deviceID)
-	// also notify v3 side so it can remove the connection from ConnMap
+func (h *Handler) OnExpiredToken(accessTokenHash, userID, deviceID string) {
+	h.v2Store.TokensTable.Delete(accessTokenHash)
+	// Notify v3 side so it can remove the connection from ConnMap
 	h.v2Pub.Notify(pubsub.ChanV2, &pubsub.V2ExpiredToken{
+		UserID:   userID,
 		DeviceID: deviceID,
 	})
```
Comment (DMRobertson) [inline sync2/handler2/handler.go:163]: At first I didn't delete the token, but then the token would get selected on incoming requests only for the homeserver to reject it. Loop of doom.

Do we care about deleting messages etc? We could have some kind of cleanup that nukes messages for devices if they've had no access tokens for at least e.g. 30 days.

File: sync2/poller.go [hunk 3]
```diff
@@ -151,15 +156,21 @@ func (h *PollerMap) NumPollers() (count int) {
...
 		go h.execute()
 	}
-	poller, ok := h.Pollers[deviceID]
+	poller, ok := h.Pollers[pid]
 	// a poller exists and hasn't been terminated so we don't need to do anything
 	if ok && !poller.terminated.Load() {
+		if poller.accessToken != accessToken {
+			logger.Warn().
+				Str("user_id", pid.UserID).
+				Str("device_id", pid.DeviceID).
```
Comment (DMRobertson) [inline sync2/poller.go:171]: Redundant, this is set on the logger by the caller

File: sync2/poller.go [hunk 4]
```diff
@@ -169,28 +180,31 @@ func (h *PollerMap) EnsurePolling(accessToken, userID, deviceID, v2since string,
...
 	go poller.Poll(v2since)
-	h.Pollers[deviceID] = poller
+	h.Pollers[pid] = poller
 
 	h.pollerMu.Unlock()
 	if needToWait {
 		poller.WaitUntilInitialSync()
 	} else {
-		logger.Info().Str("user", userID).Msg("a poller exists for this user; not waiting for this device to do an initial sync")
+		logger.Info().Str("user", poller.userID).Msg("a poller exists for this user; not waiting for this device to do an initial sync")
```
Comment (DMRobertson) [inline sync2/poller.go:207]: `user` field set by caller

File: sync2/tokens_table.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+			"token (hash %s) found with user+device mismatch: got (%s, %s), expected (%s, %s)",
+			tokenHash, gotUserID, gotDeviceID, userID, deviceID,
+		)
+	}
+	accessToken, err = t.decrypt(encToken)
+	return
+}
+
+// Delete looks up a token by its hash and deletes the row.
+// TODO: what happens if the hash isn't found?
```
Comment (DMRobertson) [inline sync2/tokens_table.go:226]: Probably return NoRowsError and log as a warning.

File: sync2/tokens_table_test.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package sync2
+
+import (
+	"testing"
+)
+
+// Sanity check that different tokens have different hashes
```
Comment (DMRobertson) [inline sync2/tokens_table_test.go:7]: Porting this from `request_test.go`

File: sync3/connmap.go [hunk 7]
```diff
@@ -71,7 +71,7 @@ func (m *ConnMap) CreateConn(cid ConnID, newConnHandler func() ConnHandler) (*Co
 	conn = NewConn(cid, h)
 	m.cache.Set(cid.String(), conn)
 	m.connIDToConn[cid.String()] = conn
```
Comment (DMRobertson) [inline sync3/connmap.go:73]: maybe this should now be a map from ConnID structs, not strings?

File: sync3/handler/handler.go [hunk 8]
```diff
@@ -276,25 +281,53 @@ func (h *SyncLiveHandler) serve(w http.ResponseWriter, req *http.Request) error
...
+			}
+		}
+	}
+	log := hlog.FromRequest(req).With().Str("user", token.UserID).Str("device", token.DeviceID).Logger()
+
+	// Record the fact that we've recieved a request from this token
+	err = h.V2Store.TokensTable.MaybeUpdateLastSeen(token, time.Now())
+	if err != nil {
+		// Not fatal---log and continue.
+		log.Warn().Msg("Unable to update last seen timestamp")
```
Comment (DMRobertson) [inline sync3/handler/handler.go:320]: Include the `err` here

File: tests-integration/connection_test.go [hunk 9]
```diff
@@ -639,17 +639,10 @@ func TestExpiredAccessToken(t *testing.T) {
 	})
 	// now expire the token
 	v2.invalidateToken(aliceToken)
-	// now do another request, this should 400 as it expires the session
+	// now do another request, this should 401
 	req := sync3.Request{}
 	req.SetTimeoutMSecs(1)
 	_, body, statusCode := v3.doV3Request(t, context.Background(), aliceToken, res.Pos, req)
```
Comment (DMRobertson) [inline tests-integration/connection_test.go:645]: Wasn't sure if this change in behaviour is desirable?

File: tests-integration/v3_test.go [hunk 10]
```diff
@@ -97,6 +110,12 @@ func (s *testV2Server) userID(token string) string {
 	return s.tokenToUser[token]
 }
 
+func (s *testV2Server) deviceID(token string) string {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.tokenToDevice[token]
+}
```
Comment (DMRobertson) [inline tests-integration/v3_test.go:117]: I wonder if this is safe. Should the mutex guard the (user, device) pair?

File: sync2/tokens_table.go (hunk 5 above)
Comment (kegsay) [reply sync2/tokens_table.go:226]: It's not great to return errors when nothing was deleted, as it isn't a fatal condition and is more likely to cause issues as people forget to check `err` for this specific type of error.

File: sync2/tokens_table.go [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+func (t *TokensTable) GetTokenAndSince(userID, deviceID, tokenHash string) (accessToken, since string, err error) {
+	var encToken, gotUserID, gotDeviceID string
+	row := t.db.QueryRow(`
+		SELECT token_encrypted, since, user_id, device_id
+		FROM syncv3_sync2_tokens JOIN syncv3_sync2_devices USING (user_id, device_id)
+		WHERE token_hash = $1;
+	`, tokenHash)
+	err = row.Scan(&encToken, &since, &gotUserID, &gotDeviceID)
```
Comment (kegsay) [inline sync2/tokens_table.go:211]: You'd typically inline this e.g `if err := QueryRow(...).Scan(&foo); err != nil { ... }`.

File: sync2/tokens_table.go [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+	if err != nil {
+		return
+	}
+	if gotUserID != userID || gotDeviceID != deviceID {
```
Comment (kegsay) [inline sync2/tokens_table.go:215]: We definitely need the inserts in `syncv3_sync2_tokens` and `syncv3_sync2_devices` to be in a txn then if we are going to return an error if they aren't set.

File: sync2/tokens_table.go [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+	)
+	if err != nil {
+		return nil, err
+	}
+	token.AccessToken = plaintextToken
+	token.AccessTokenHash = tokenHash
+	return &token, nil
+}
+
+type TokenWithSince struct {
```
Comment (kegsay) [inline sync2/tokens_table.go:128]: Bad name as it is awkward when we want to add additional fields to this struct beyond "since".

**Review by kegsay — APPROVED** [review]
Good so far; please add unit tests for new tables.

**Comment by DMRobertson** [comment]
> please add unit tests for new tables.

There were some, but I've pulled some of this out to a separate file in 5404893.

File: sync3/handler/handler.go [hunk 14]
```diff
@@ -366,19 +367,58 @@ func (h *SyncLiveHandler) setupConnection(req *http.Request, syncReq *sync3.Requ
...
+			}
+		}
+		log.Warn().Err(err).Msg("failed to get user ID from device ID")
+		return nil, &internal.HandlerError{
+			StatusCode: http.StatusBadGateway,
+			Err:        err,
+		}
+	}
+
+	// TODO: should the two inserts be wrapped in a transaction?
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply sync3/handler/handler.go]: Done in 5c9afec.

File: sync2/tokens_table.go (hunk 5 above)
Comment (DMRobertson) [reply sync2/tokens_table.go]: Thanks. I've made 14cf838 log a warning but otherwise ignored this scenario.

File: sync2/tokens_table.go (hunk 13 above)
Comment (DMRobertson) [reply sync2/tokens_table.go]: I wasn't happy with it (but also I find it relatively easy to rename things---IntelliJ's/Goland's refactoring tools seem to work very well for go.

**Comment by DMRobertson** [comment]
This seems happy other than #36. Going to merge into the oidc branch.

**Pull request merged**